Cms/new recurring background jobs (#5630)
umbraco/UmbracoDocs: 13/umbraco-cms/reference/scheduling.md

{
    public TimeSpan Period { get => TimeSpan.FromMinutes(60); }

    // Runs on all servers
    public ServerRole[] ServerRoles { get => Enum.GetValues<ServerRole>(); }

    // No-op event as the period never changes on this job
    public event EventHandler PeriodChanged { add { } remove { } }

    private readonly IContentService _contentService;
    private readonly ICoreScopeProvider _scopeProvider;

    public CleanUpYourRoom(
        IContentService contentService,
        ICoreScopeProvider scopeProvider)
    {
        _contentService = contentService;
        _scopeProvider = scopeProvider;
    }

    public Task RunJobAsync()
    {
        // Wrap the three content service calls in a scope to do it all in one transaction.
        using ICoreScope scope = _scopeProvider.CreateCoreScope();

        int numberOfThingsInBin = _contentService.CountChildren(Constants.System.RecycleBinContent);

        if (_contentService.RecycleBinSmells())
        {
            _contentService.EmptyRecycleBin(userId: -1);
        }
        // Remember to complete the scope when done.
        scope.Complete();
        return Task.CompletedTask;
    }
}
```

## Complex example

> The complex example builds on the previous one by injecting additional services, including a logger to log error messages, a profiler to capture timings, and an `IServerRoleAccessor` to log the current server role. Additionally it injects an `IOptionsMonitor` to allow the period to be updated while the server is running, and demonstrates how to trigger the `PeriodChanged` event to signal the job's host.

Reviewer: ```suggestion
The complex example builds on the previous one by injecting additional services. It includes a logger to log error messages, a profiler to capture timings, and an `IServerRoleAccessor` to log the current server role.  Additionally, it injects an `IOptionsMonitor` to allow the period to be updated while the server is running. It also demonstrates how to trigger the `PeriodChanged` event to signal the job's host.
```

The following notifications are available:

-   Starting
-   Started
-   Stopping
-   Stopped
-   Executing
-   Executed
-   Failed
-   Ignored

### Start / Stop

The Starting/Started and Stopping/Stopped notification pairs are published when the `RecurringBackgroundJobHostedService` is started or stopped. The start event normally occurs soon after application start as part of the .Net WebHost startup process. Similarly the stop event would happen as part of application shutdown.

These notifications are there to support low-level debugging of background jobs to ensure they are starting / stopping correctly. Due to the timing of the notification, all handlers associated with these notifications should not depend on any Umbraco services, including database access.

### Ignored

The Ignored notification is published when a background job's scheduled is triggered, but of the Umbraco runtime checks prevents it from running.

> This notification is there to support low-level debugging of background jobs to ascertain why they are / are not running. As the runtime checks include runtime state readiness, this event may be triggered during the install phase. Any notification handlers associated with this notification should **also** conduct their own checks before relying on Umbraco services, including database access.

Reviewer: ```suggestion
This notification is there to support low-level debugging of background jobs to ascertain why they are/aren't running. As the runtime checks include runtime state readiness, this event may be triggered during the install phase. Any notification handlers associated with this notification should **also** conduct their own checks before relying on Umbraco services, including database access.
```

> ### Executing / Executed / Failed

Reviewer: ```suggestion
### Executing/Executed/Failed
```

> These notifications will be triggered in pairs depending on the success / failure of the job itself. The executing notification is triggered before the job is run. The executed notification is triggered after the job completes. The failed notification is triggered from the catch block if an exception is thrown.

Reviewer: ```suggestion
These notifications will be triggered in pairs depending on the success/failure of the job itself. 
- The executing notification is triggered before the job is run. 
- The executed notification is triggered after the job is completed. 
- The failed notification is triggered from the catch block if an exception is thrown.
```
